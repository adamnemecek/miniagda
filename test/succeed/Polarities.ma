-- 2010-06-19, 2010-11-09

let Const : ++ Set -> . Set -> Set 
          = \ A -> \ X -> A

let DNeg : ^ Set -> + Set -> Set
         = \ B -> \ A -> * (* A -> B) -> B

data Empty : Set {}

sized data Nat : + Size -> Set
{ zero : [i : Size] -> Nat ($ i)
; succ : [i : Size] -> ^ Nat i -> Nat ($ i)
}

let Cont' : + Set -> Set
         = DNeg Empty

-- the following holds already because of whnf computation
let cast' : [i : Size] -> ^ Cont' (Nat i) -> Cont' (Nat #)
         = \ i -> \ x -> x

data Cont +(A : Set) : Set 
{ cont : (uncont : DNeg Empty A) -> Cont A
}

-- the following holds because Cont is a datatype (pol. already impl.)
let cast : [i : Size] -> ^ Cont (Nat i) -> Cont (Nat #)
         = \ i -> \ x -> x

-- hide positivity behind recursion
fun Id : * Nat # -> ++Set -> Set
{ Id (zero .#)   A = A
; Id (succ .# n) A = A
}

let kast : [i : Size] -> [n : Nat i] -> Id n (Nat i) -> Id n (Nat #)
         = \ i -> \ n -> \ x -> x

data Tree -(B : Set) ++(A : Set) : Set
{ leaf : Tree B A
; node : A -> (B -> Tree B A) -> Tree B A
}

sized data STree -(B : Set) ++(A : Set) : +Size -> Set
{ sleaf : [i : Size] -> STree B A ($ i)
; snode : [i : Size] -> A -> (B -> STree B A i) -> STree B A ($ i)
}

data Mu ++(F : ++Set -> Set) : Set
{ inn : F (Mu F) -> Mu F
}

{-
  .(p)  = o
  ++(p) = p
  +(++) = +
  +(p)  = p
  -(++) = -
  -(+)  = -
  -(-)  = +
  -(p)  = p 
  o(o)  = o
  o(++) = .
  o(+)  = .
  o(-)  = .
  o(.)  = .

  -(Gamma) |- A : s  Gamma |- B : s
  ---------------------------------
  Gamma |- A -> B : s

  -(Gamma) |- A : s  Gamma, x : A |- B : s
  ----------------------------------------
  Gamma |- p(x : A) -> B : s

  --------------------------------  p in {++,+,o}
  Gamma, p(x : A), Gamma' |- x : A
  
  Gamma, p(x : A) |- t : B
  ----------------------------
  Gamma |- \xt : p(x : A) -> B

  Gamma |- r : p(x : A) -> B   p(Gamma) |- s : A
  ----------------------------------------------
  Gamma |- r s: B[s/x]
    

-}