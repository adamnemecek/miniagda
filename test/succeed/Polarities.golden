--- opening "Polarities.ma" ---
--- scope checking ---
--- type checking ---
type  Const : ++ Set -> .[Set] -> Set
type  Const = \ A -> [\ X ->] A
type  DNeg : ^ Set -> + Set -> Set
type  DNeg = \ B -> \ A -> (A -> B) -> B
type  Empty : Set
type  Nat : + Size -> Set
term  Nat.zero : .[s!ze : Size] -> .[i < s!ze] -> Nat s!ze
term  Nat.zero : .[i : Size] -> < Nat.zero i : Nat $i >
term  Nat.succ : .[s!ze : Size] -> .[i < s!ze] -> ^ Nat i -> Nat s!ze
term  Nat.succ : .[i : Size] -> ^(y1 : Nat i) -> < Nat.succ i y1 : Nat $i >
type  Cont' : + Set -> Set
type  Cont' = DNeg Empty
term  cast' : .[i : Size] -> ^ Cont' (Nat i) -> Cont' (Nat #)
term  cast' = [\ i ->] \ x -> x
type  Cont : +(A : Set) -> Set
term  Cont.cont : .[A : Set] -> ^(uncont : DNeg Empty A) -> < Cont.cont uncont : Cont A >
term  uncont : .[A : Set] -> (cont : Cont A) -> DNeg Empty A
{ uncont [A] (Cont.cont #uncont) = #uncont
}
term  cast : .[i : Size] -> ^ Cont (Nat i) -> Cont (Nat #)
term  cast = [\ i ->] \ x -> x
type  Id : Nat # -> ++ Set -> Set
{ Id (Nat.zero [.#]) A = A
; Id (Nat.succ [.#] n) A = A
}
term  kast : .[i : Size] -> .[n : Nat i] -> Id n (Nat i) -> Id n (Nat #)
term  kast = [\ i ->] [\ n ->] \ x -> x
type  Tree : -(B : Set) -> ++(A : Set) -> Set
term  Tree.leaf : .[B : Set] -> .[A : Set] -> < Tree.leaf : Tree B A >
term  Tree.node : .[B : Set] -> .[A : Set] -> ^(y0 : A) -> ^(y1 : B -> Tree B A) -> < Tree.node y0 y1 : Tree B A >
type  STree : -(B : Set) -> ++(A : Set) -> + Size -> Set
term  STree.sleaf : .[B : Set] -> .[A : Set] -> .[s!ze : Size] -> .[i < s!ze] -> STree B A s!ze
term  STree.sleaf : .[B : Set] -> .[A : Set] -> .[i : Size] -> < STree.sleaf i : STree B A $i >
term  STree.snode : .[B : Set] -> .[A : Set] -> .[s!ze : Size] -> .[i < s!ze] -> ^ A -> ^ (B -> STree B A i) -> STree B A s!ze
term  STree.snode : .[B : Set] -> .[A : Set] -> .[i : Size] -> ^(y1 : A) -> ^(y2 : B -> STree B A i) -> < STree.snode i y1 y2 : STree B A $i >
type  Mu : ++(F : ++ Set -> Set) -> Set
term  Mu.inn : .[F : ++ Set -> Set] -> ^(y0 : F (Mu F)) -> < Mu.inn y0 : Mu F >
--- evaluating ---
--- closing "Polarities.ma" ---
